### ISSUE TITLE
Best practice: amp-video-docking: unexpected keyboard behaviour

### ISSUE THREAD

**Issue by TetraLogicalHelpdesk** [ampproject/amphtml#31163]
## WCAG Level
Level A

## Priority
High

## Type
Navigation

## Pages/screens/components affected
* [amp-video-docking](https://amp.dev/documentation/components/amp-video-docking/)

## Description
The current implementation of `amp-video-docking` presents a variety of challenges for keyboard users.

While the video itself can be triggered via the keyboard, its visual "docking" behavior appears to only happen on scrolling (rather than when it's determined that the original video element is not in the visible viewport anymore). When navigating using `Tab`/`Shift`+`Tab`, the video does not "dock" when the user's keyboard focus change results in the original video element landing outside of the visible viewport.

![The docked/floating video](https://user-images.githubusercontent.com/54933393/99181446-92332d00-2726-11eb-9b65-f96aebf8026f.png)

Once the video is docked, its regular `<video>` element controls are suppressed. In their place, custom controls appear when the docked video is hovered with the mouse. As a result, there are no keyboard-focusable controls left for a keyboard user to navigate to. Once the video is docked, keyboard users have to manually scroll the page (using the cursor keys) until they reach the original video's location - at which point the docked video will "undock" itself and return to its previous position in the page, and can then be operated. Again, if the users navigate using `Tab`/`Shift`+`Tab`, the video will not "undock" itself even if its original location in the page is visible.

## User impact
While technically this component passes WCAG 2.1 Success Criterion 2.1.1 Keyboard, as there is *a way* for keyboard users to still operate the video, the experience is nonetheless confusing and unnecessarily convoluted.

## Required solution
Due to the unusual nature of this component, there is no clear "best" solution for this issue. At a high level, we would recommend:

* not relying on scrolling to determine when the video should be docked/undocked, but instead checking programmatically if the element is visible in the viewport (for instance, using the [Intersection Observer API](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API), which will work even if the element was moved out of view as a result of a keyboard navigation using `Tab`/`Shift`+`Tab`).
* leaving at least a single keyboard-focusable element in place of where the video was originally located in the page; this way, keyboard users will be able to `Tab` to the location - at which point the video should be undocked, and focus programmatically moved to the newly reinstated `<video>` with its default controls.

This solution must be applied to all instances of the issue identified within the test sample, then applied to all other instances of the same issue identified throughout the rest of the website.

## Test procedure(s)
Use these steps to confirm that the solution has been correctly applied to issues identified within the test sample, and to test the rest of the website for instances of the same issue:
1. Using the keyboard, navigate to the video (using `Tab`/`Shift`+`Tab`) and begin playback
2. Continue navigating using the keyboard - in particular, moving to other focusable elements (using `Tab`/`Shift`+`Tab`)
3. Verify that the video is docked when the video element is out of the viewport
4. Once docked, return to the original location of the video in the page (using `Tab`/`Shift`+`Tab`)
5. Verify that the original location of the video can be reached, and the video is undocked/returned to its normal location in the page, and displays its regular native `<video>` element controls

### Definition of done
Complete all of these tasks before closing this issue or indicating it is ready for retest:
* All issues identified within the test sample have been resolved.
* The rest of the website has been tested for the same issue.
* All issues identified throughout the rest of the website have been resolved or filed as new issues.

## Related standards
* [WCAG 2.1 Success Criterion 2.1.1 Keyboard (Level A)](https://www.w3.org/TR/WCAG21/#keyboard)

## More information
* [WebAIM: Keyboard](https://webaim.org/techniques/keyboard/)

## Test data
Test date: November 2020
